sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -331,6 +331,8 @@ def trim_links(links, link_count_limit, link_attribute_count_limit) # rubocop:di
           # Fast path (likely) common cases.
           return nil if links.nil?
 
+          links.filter! { |link| link.span_context.valid? }
```

Comment:
Technically, we shouldn't modify the `links` array. It is handed to us by the caller (well, a few levels up the stack) and they could modify it subsequently. On the fast(er) path, we check the content and either use it as is (if already frozen) or clone & freeze the array. On the slow path, we'll create a new array before modifying its contents.

All of which to say, I think we should not `filter!` here - instead we should validate on the fast(er) path below (I'll show what I mean) and then `filter` before `last` on the slow path.